extension/src/components/UpdatesContainer.tsx

```diff
@@ -1,11 +1,48 @@
 import { useRef } from "preact/hooks";
-import { updatesStore } from "../models/UpdatesModel";
+import { updatesStore, type UpdateTreeNode } from "../models/UpdatesModel";
 import { UpdateTreeNodeComponent } from "./UpdateTreeNode";
 import { useSignalEffect } from "@preact/signals";
+import { settingsStore } from "../models/SettingsModel";
+
+const nodesAreEqual = (a: UpdateTreeNode, b: UpdateTreeNode): boolean => {
+	return (
+		a.update.signalId === b.update.signalId &&
+		a.children.length === b.children.length &&
+		a.children.every((child, index) => nodesAreEqual(child, b.children[index]))
+	);
+};
+
+interface CollapsedTree {
+	tree: UpdateTreeNode[];
+	counts: WeakMap<UpdateTreeNode, number>;
+}
+
+const collapseTree = (nodes: UpdateTreeNode[]): CollapsedTree => {
+	const tree: UpdateTreeNode[] = [];
+	let lastNode: UpdateTreeNode | null = null;
+	const counts = new WeakMap<UpdateTreeNode, number>();
+
+	for (const node of nodes) {
+		if (lastNode && nodesAreEqual(lastNode, node)) {
+			// If the current node is equal to the last one, skip it
+			counts.set(lastNode, (counts.get(lastNode) ?? 1) + 1);
+			continue;
+		}
+		tree.push(node);
+		lastNode = node;
+	}
+
+	return { tree, counts };
+};
 
 export function UpdatesContainer() {
 	const updatesListRef = useRef<HTMLDivElement>(null);
 	const updateTree = updatesStore.updateTree.value;
+	let collapsedTree = updateTree;
+	let counts: WeakMap<UpdateTreeNode, number> | undefined;
+	if (settingsStore.settings.grouped) {
```

Comment:
Let's move this to the `updatesStore`, it could be a computed that only fires when this setting is on. We could also when we collapse change the structure of the tree a bit where `count` is a property

Reply:
We could replace updateTree and give it a type property so you can return two different shapes. Whatever works best for you

Reply:
you can see a rough attempt at this in 69f4641